src/backend/access/heap/heapam.c

```diff
@@ -5864,6 +5875,7 @@ heap_finish_speculative(Relation relation, ItemPointer tid)
 
 	buffer = ReadBuffer(relation, ItemPointerGetBlockNumber(tid));
 	LockBuffer(buffer, BUFFER_LOCK_EXCLUSIVE);
+	ReleaseBuffer(buffer); /* NEON: release buffer pinned by heap_insert */
```

Comment:
This feels like a pretty surprising location for this. I'd suggest putting it at the end of the function, after UnlockReleaseBuffer(buffer). Same in heap_abort_speculative.